internal/pkg/fleetapi/uploader/uploader.go

```diff
@@ -0,0 +1,223 @@
+			tr = bytes.NewReader(b.Bytes())
+			r.wait.Wait()
+			continue
+		}
+		return resp, err
+	}
+	return resp, err
+}
+
+// URI calls the underlying Sender's URI method.
+func (r *retrySender) URI() string {
+	return r.c.URI()
+}
+
+// Client provides methods to upload a file to ES through fleet-server.
+type Client struct {
+	agentID string
+	c       client.Sender
+}
+
+// New returns a new Client for the agent identified by the passed id.
+// The sender is wrapped with retry logic specified by the Uploader config.
+// Any request that would return a 429 (too many requests) is retried (up to maxRetries times) with a backoff.
+func New(id string, c client.Sender, cfg config.Uploader) *Client {
+	return &Client{
+		agentID: id,
+		c: &retrySender{
+			c:    c,
+			max:  cfg.MaxRetries,
+			wait: backoff.NewEqualJitterBackoff(nil, cfg.InitDur, cfg.MaxDur),
+		},
+	}
+}
+
+// New sends a new file upload request to the fleet-server.
+func (c *Client) New(ctx context.Context, r *NewUploadRequest) (*NewUploadResponse, error) {
+	b, err := json.Marshal(r)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := c.c.Send(ctx, http.MethodPost, PathNewUpload, nil, nil, bytes.NewBuffer(b))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, client.ExtractError(resp.Body)
+	}
+
+	var newUploadResp NewUploadResponse
+	dec := json.NewDecoder(resp.Body)
+	if err := dec.Decode(&newUploadResp); err != nil {
+		return nil, err
+	}
+	return &newUploadResp, nil
+}
+
+// Chunk uploads a file chunk to fleet-server.
+func (c *Client) Chunk(ctx context.Context, uploadID string, chunkID int, sha256Hash []byte, r io.Reader) error {
+	h := http.Header{"X-Chunk-Sha2": {fmt.Sprintf("%x", sha256Hash)}}
+	resp, err := c.c.Send(ctx, "PUT", fmt.Sprintf(PathChunk, uploadID, chunkID), nil, h, r)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return client.ExtractError(resp.Body)
+	}
+	return nil
+}
+
+// Finish calls the finalize endpoint for the file upload.
+func (c *Client) Finish(ctx context.Context, id string, r *FinishRequest) error {
+	b, err := json.Marshal(r)
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.c.Send(ctx, "POST", fmt.Sprintf(PathFinishUpload, id), nil, nil, bytes.NewBuffer(b))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return client.ExtractError(resp.Body)
+	}
+	return nil
+}
+
+// UploadDiagnostics is a wrapper to upload a diagnostics request identified by the passed action id contained in the buffer to fleet-server.
+//
+// A buffer is used instead of a reader as we need to know the file length before uploading.
+func (c *Client) UploadDiagnostics(ctx context.Context, id string, b *bytes.Buffer) (string, error) {
+	size := b.Len()
+	upReq := NewUploadRequest{
+		ActionID: id,
+		AgentID:  c.agentID,
+		Source:   "elastic-agent",
```

Comment:
I believe this needs to be `"agent"` to match specific patterns for the index template names